zio-http/src/main/scala/zio/http/package.scala

```diff
@@ -1,17 +1,20 @@
 package zio
 
-import zio.http.socket.WebSocketChannelEvent
 import zio.stacktracer.TracingImplicits.disableAutoTrace // scalafix:ok;
 
 package object http extends PathSyntax with RequestSyntax with RouteDecoderModule {
-  type HttpApp[-R, +E] = Http[R, E, Request, Response]
-  type UHttpApp        = HttpApp[Any, Nothing]
-  type RHttpApp[-R]    = HttpApp[R, Throwable]
-  type EHttpApp        = HttpApp[Any, Throwable]
-  type UHttp[-A, +B]   = Http[Any, Nothing, A, B]
 
-  type ResponseZIO[-R, +E]                   = ZIO[R, E, Response]
-  type UMiddleware[+AIn, -BIn, -AOut, +BOut] = Middleware[Any, Nothing, AIn, BIn, AOut, BOut]
+  type RequestHandler[-R, +Err] = Handler[R, Err, Request, Response]
+  type AppHandler[-R]           = RequestHandler[R, Response]
+
+  type RequestHandlerMiddleware[-R, +Err] = HandlerAspect[R, Err, Request, Response, Request, Response]
```

Comment:
I suggest, maybe, that we define a new trait called `Middleware[R, Err]`, and have it extend `MiddlewareAspect[R, Err, Request, Response, Request, Response]`. Then we'll have a reasonable place to put all the `Middleware` (into the companion object of `Middleware`).

Reply:
I'm not sure I understand this suggestion. 

- if I understand correctly `MiddlewareAspect` is what above defined as `RequestHandlerMiddleware` - but in other cases we used `XYAspect` as something that transforms `XY` so that name does not feel right to me
- What is the reason for making it a trait?
- Currently there are a couple of traits defining all kind of middlewares (ones working on handlers, also some working on routes): `RequestHandlerMiddlewares`, `HttpRouteMiddlewares`, `Cors`, `Metrics`, etc. And there is a single`Middleware` object merging all these to make them easier to find. What subset would go into `Middleware` by this suggestion, and where would the remaining ones go?

Reply:
I am not 100% sure, but it would be nice to retain the terminology of middleware because it's well-established, whereas aspect terminology, however applicable and consistent, is less familiar and less domain-specific.

In addition, having a `trait` `Middleware` improves organization: because we are always teaching Scala developers to look into the companion object of `XYZ` to find constructors for `XYZ`.

Using `Middleware` as a "package", which is what we are doing in this pull request, helps to organize the middlewares, but not in any way that is discoverable; and in a way that is quite different than the typical type/companion object relationship.

Now, you mention we have some handler middleware, and route middleware. 

Well, that's unfortunate, because the `Middleware` companion object to a `Middleware` type (which extends `HandlerAspect`) should only contain handler middleware. Not route middleware.

But if there are only a couple route middleware, then maybe we can just implement them as methods on route. When I looked, I saw only really 1 route aspect ("route middleware") which is a sign the majority of what a user would think of as "middleware" moved to handler middleware. If true, then the middleware name could be reused just for route middleware, and that would also result in less code breakage.

Reply:
There is not only one route middleware, but much less than handler ones. Basically those middlewares that somehow extend or limit the set of handled requests, or needs to observe the event of not handling one:


- `metrics`, to be able to count unhandled cases
- `CORS` because it adds new routes (handles `OPTIONS`)
- `dropTrailingSlash` because it "doubles" the allowed routes
- `allow` and `allowZIO` because they dynamically change the routing

Reply:
I was thinking of also exposing each "kind" of middleware in a companion object of the corresponding type alias for example `RequestHandlerMiddleware` (by pulling in the same underlying trait that I mixed in `Middleware` now).

Not sure if that would help or make it even worse :)